### ISSUE TITLE
[C++] xthread: Failed to specialize function template 'unknown-type std::invoke(_Callable &&,_Types &&...) noexcept(<expr>)'

### ISSUE THREAD

**Issue by vutuyetmai130398** [hassanyf/Simple-Server-and-Chat-Program#6]
`#include "stdafx.h"
#include "ServerType.h"

bool CheckResult(int iResult) {
	if (iResult) {
		cout << "WSAStartup failed with error: " << iResult<<endl;
		return 0;
	}
	return 1;
}

ServerType::ServerType()
{
	this->iResult = WSAStartup(MAKEWORD(2, 2), &this->wsaData);
	this->addrin.result = NULL;
	this->recvbuf = new char[DEFAULT_BUFLEN];
	this->recvbuflen = DEFAULT_BUFLEN;
	this->_client = INVALID_SOCKET;
	this->_server = INVALID_SOCKET;
}

bool ServerType::InitWinSock()
{
	if (CheckResult(this->iResult)) {
		ZeroMemory(&this->addrin.hints, sizeof(this->addrin.hints));
		this->addrin.hints.ai_family = AF_INET;
		this->addrin.hints.ai_socktype = SOCK_STREAM;
		this->addrin.hints.ai_protocol = IPPROTO_TCP;
		this->addrin.hints.ai_flags = AI_PASSIVE;

		this->iResult = getaddrinfo(NULL, DEFAULT_PORT, &this->addrin.hints, &this->addrin.result);
		if (this->iResult) {
			cout << "Getaddrinfo failed with error: " << this->iResult<<endl;
			WSACleanup();
			return 0;
		}

		return 1;
	}
	return false;
}

bool ServerType::ConnectServer()
{
	//Create a SOCKET for connecting server
	this->_server = socket(this->addrin.result->ai_family, this->addrin.result->ai_socktype, this->addrin.result->ai_protocol);

	if (this->_server == INVALID_SOCKET) {
		cout << "Socket failed with error: " << WSAGetLastError() << endl;
		freeaddrinfo(this->addrin.result);
		WSACleanup();
		return 0;
	}

	cout << "Binding socket ..." << endl;
	bind(this->_server, this->addrin.result->ai_addr, (int)this->addrin.result->ai_addrlen);
	
	freeaddrinfo(this->addrin.result);

	thread thdReceive(&ServerType::ListenClient, this);
	thdReceive.detach();
	return false;
}

void ServerType::ListenClient()
{
	try
	{
		while (1) {
			listen(this->_server, SOMAXCONN);
			SOCKET client = INVALID_SOCKET;
			client = accept(this->_server, NULL, NULL);

			if (client == INVALID_SOCKET) {
				continue;
			}
			else {
				this->_clientList.push_back(client);
				thread ReceiveThr(&ServerType::ReceiveMsg, std::ref(client));
				ReceiveThr.detach();
			}
		}
	}
	catch (const std::exception&)
	{

	}
}

void ServerType::ReceiveMsg(SOCKET client)
{
	string msg = "";

	while (true)
	{
		memset(this->recvbuf, 0, DEFAULT_BUFLEN);

		if (client != INVALID_SOCKET) {
			this->iResult = recv(client, this->recvbuf, DEFAULT_BUFLEN, 0);

			if (this->iResult != SOCKET_ERROR) {
				if (strcmp("", this->recvbuf)) {
					msg = this->recvbuf;
					cout << msg << endl;
					for (std::vector<SOCKET>::iterator it = this->_clientList.begin(); it != this->_clientList.end(); it++) {
						if (*it != INVALID_SOCKET && *it!=client) {
							this->iResult = send(*it, msg.c_str(), strlen(msg.c_str()), 0);
						}
					}
				}
			}
			else
			{
				msg = "Client Disconnect!!\n";
				cout << msg;
				for (std::vector<SOCKET>::iterator it = this->_clientList.begin(); it != this->_clientList.end(); it++) {
					if (*it == client) {
						closesocket(*it);
					}
					if (*it != INVALID_SOCKET) {
						this->iResult = send(*it, msg.c_str(), strlen(msg.c_str()), 0);
					}
				}
				closesocket(client);
			}
			break;
		}
	}
	
}

ServerType::~ServerType()
{
}
`
i have a project which is used "Chat app" written by C++. But i have some problems in the xthread file, I cannot fix them, I use "this" in the "thread" function, but it not successful. 
Please help me!

**Comment by DGPickett** [comment]
thread ReceiveThr(&ServerType::ReceiveMsg, std::ref(client)); 
thread thdReceive(&ServerType::ListenClient, this);

These calls pass a pointer to a pointer to a function/method, but what is needed is just the function/method name.  The declaration of the thread constructor has first argument void name( void * arg_name ), so pass by reference is a given, as subroutine names are pointers to the subroutine code.  See the online examples: http://www.cplusplus.com/reference/thread/thread/thread/
